pkg/l2announcer/l2announcer.go

```diff
@@ -0,0 +1,1113 @@
+	L2AnnounceTable      statedb.Table[*tables.L2AnnounceEntry]
+	StateDB              statedb.DB
+	JobRegistry          job.Registry
+}
+
+// L2Announcer takes all L2 announcement policies and filters down to those that match the labels of the local node. It
+// then searches all services that match the selectors of the policies. For each service, we attempt to take a lease,
+// the holder node persists all IPs and netdev combinations selected by the policy to the L2AnnounceTable. Datapath
+// components consume them and handle traffic for the IP+netdev entries.
+type L2Announcer struct {
+	params l2AnnouncerParams
+
+	svcStore    resource.Store[*slim_corev1.Service]
+	policyStore resource.Store[*cilium_api_v2alpha1.CiliumL2AnnouncementPolicy]
+	localNode   *v2.CiliumNode
+
+	jobgroup job.Group
+
+	leaderChannel chan leaderElectionEvent
+	devicesChan   chan []string
+
+	// selectedPolicies matching the current node.
+	selectedPolicies map[resource.Key]*selectedPolicy
+	// Services which are selected by one or more policies for which we thus want to participate in leader election.
+	// Indexed by service key.
+	selectedServices map[resource.Key]*selectedService
+	// A list of devices which can be matched by the policies
+	devices []string
+}
+
+func NewL2Announcer(params l2AnnouncerParams) *L2Announcer {
+	// These values were picked because it seemed right, change if necessary
+	const leaderElectionBufferSize = 16
+	const devicesBufferSize = 16
+	announcer := &L2Announcer{
+		params: params,
+		jobgroup: params.JobRegistry.NewGroup(
+			job.WithLogger(params.Logger),
+			job.WithPprofLabels(pprof.Labels("cell", "l2-announcer")),
+		),
+		selectedServices: make(map[resource.Key]*selectedService),
+		selectedPolicies: make(map[resource.Key]*selectedPolicy),
+		leaderChannel:    make(chan leaderElectionEvent, leaderElectionBufferSize),
+		devicesChan:      make(chan []string, devicesBufferSize),
+	}
+
+	// Can't operate or GC if client set is disabled
+	if !params.Clientset.IsEnabled() {
+		return announcer
+	}
+
+	params.Lifecycle.Append(announcer.jobgroup)
+
+	if !params.DaemonConfig.EnableL2Announcements {
+		// If the L2 announcement feature is disabled, garbage collect any leases from previous runs when the feature
+		// might have been active. Just once, not on a timer.
+		announcer.jobgroup.Add(job.OneShot("l2-announcer lease-gc", announcer.leaseGC))
+		return announcer
+	}
+
+	announcer.jobgroup.Add(job.OneShot("l2-announcer run", announcer.run))
+	announcer.jobgroup.Add(job.Timer("l2-announcer lease-gc", announcer.leaseGC, time.Minute))
+
+	return announcer
+}
+
+// DevicesChanged can be invoked by an external component responsible for discovering all available network devices to
+// inform this component of all the devices we can use for L2 announcements.
+func (l2a *L2Announcer) DevicesChanged(devices []string) {
+	select {
+	// Send the change to the central routine to avoid race conditions
+	case l2a.devicesChan <- devices:
+	default:
```

Comment:
Fair point. We only really care about the latest update anyway. So now the function will update the slice directly which can technically cause a race-condition. But I also send a "trigger" on the channel which should take care of any issues that might have resulted from modifying the devices list halfway through. This approach is nice since it will coalesce changes if multiple happen in short succession.